vpr/src/place/initial_placement.cpp

```diff
@@ -236,101 +449,83 @@ static void initial_placement_pl_macros(int macros_max_num_tries, std::vector<st
-             * accommodate a block of that type in free_locations[itype].
-             * Choose one randomly and put blk_id there. Then we don't want to pick
-             * that location again, so remove it from the free_locations array.
-             */
-
-            auto type = pick_placement_type(logical_block, 1, free_locations);
-
-            if (type == nullptr) {
-                VPR_FATAL_ERROR(VPR_ERROR_PLACE,
-                                "Initial placement failed.\n"
-                                "Could not place block %s (#%zu); no free locations of type %s (#%d).\n",
-                                cluster_ctx.clb_nlist.block_name(blk_id).c_str(), size_t(blk_id), logical_block->name, logical_block->index);
-            }
-
-            int itype = type->index;
-
-            auto possible_sub_tiles = get_possible_sub_tile_indices(type, logical_block);
-
-            int isub_tile = get_free_sub_tile(free_locations, itype, possible_sub_tiles);
+    bool placed = false;
+    while (!placed) {
+        /* -1 is a sentinel for a non-placed block, which the code in this routine will choose a location for.
+         * If the x value is not -1, we assume something else has already placed this block and we should leave it there.
+         * For example, if the user constrained it to a certain location, the block has already been placed.
+         */
+        if (place_ctx.block_locs[blk_id].loc.x != -1) {
+            //put VTR assert
+            placed = true;
+            continue;
+        }
 
-            t_pl_loc to;
-            int ipos = vtr::irand(free_locations[itype][isub_tile] - 1);
+        auto logical_block = cluster_ctx.clb_nlist.block_type(blk_id);
 
-            to = legal_pos[itype][isub_tile][ipos];
+        t_pl_loc to;
+        PartitionRegion pr;
 
-            // Make sure that the position is EMPTY_BLOCK before placing the block down
-            VTR_ASSERT(place_ctx.grid_blocks[to.x][to.y].blocks[to.sub_tile] == EMPTY_BLOCK_ID);
+        if (is_cluster_constrained(blk_id)) {
+            pr = floorplanning_ctx.cluster_constraints[blk_id];
+        } else {
+            Region reg;
+            reg.set_region_rect(0, 0, device_ctx.grid.width() - 1, device_ctx.grid.height() - 1);
+            reg.set_sub_tile(NO_SUBTILE);
+            pr.add_to_part_region(reg);
+        }
+        bool found_legal_place = false;
+        bool floorplanning_good = false;
 
-            bool floorplan_good = cluster_floorplanning_legal(blk_id, to);
+        for (int itry = 0; itry < blocks_max_num_tries && placed == false; itry++) {
+            found_legal_place = get_legal_placement_loc(pr, to, logical_block);
 
-            if (floorplan_good) {
-                place_ctx.grid_blocks[to.x][to.y].blocks[to.sub_tile] = blk_id;
-                place_ctx.grid_blocks[to.x][to.y].usage++;
+            if (found_legal_place) {
+                floorplanning_good = cluster_floorplanning_legal(blk_id, to);
```

Comment:
Should this be done in get_legal_placement_loc? 
- Comment why this is needed (compressed_grid_approximation)
- Move this check to get_legal_placement_loc (with the comment).
- Reorder the parameters to get_legal_placement_loc (and other routines) so inputs come first, and outputs (return by reference) come after.

Reply:
no longer use this routine (place_block) because the macro and non-macro flows were merged